File: C++/Graph/797_All Paths From Source to Target.cpp
#################### DFS #############################
class Solution {
public:
    void dfs(vector<vector<int>>& graph, int u, int t, vector<vector<int>>& result,vector<int>& temp){
        temp.push_back(u);
        if(u==t){
            result.push_back(temp);
        }
        else{
            for(int &v:graph[u]){
                dfs(graph, v, t, result, temp);        
            }
        }
        
        temp.pop_back();
        
    }
    
    
    
    
    
    vector<vector<int>> allPathsSourceTarget(vector<vector<int>>& graph) {
        int n = graph.size();
        
        int source = 0;
        int target = n-1;
        vector<vector<int>> result;
        vector<int> temp;
        
        
        dfs(graph, source, target, result, temp);
        
        return result;
    }
};

################################## BFS #######################################
class Solution {
public:
vector<vector<int>> allPathsSourceTarget(vector<vector<int>>& graph) {
        int n = graph.size();
        queue<vector<int>> q;
        vector<vector<int>> ans;
    
        q.push({0});
    
    while(!q.empty()){
        vector<int> path = q.front();
        
        q.pop();
        
        if(path.back()== n-1){
            ans.push_back(path);
        }
        else
        for (auto child:graph[path.back()]){
            path.push_back(child);
            q.push(path);
            path.pop_back();
        }
    }
    return ans;        
    
    }
};

File: Python/Graph/797_All Paths From Source to Target.py
##################### DFS ######################
class Solution:
    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        def dfs(curr, path):
            if curr == len(graph)-1:
                res.append(path)
            else:
                for i in graph[curr]:
                    dfs(i, path + [i])
                    
        res= []
        
        dfs(0,[0])
        return res


##################### BFS ##########################

class Solution:
    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        ans, q = [], deque([[0]])
        while q:
            path = q.popleft()
            if path[-1] == len(graph)-1: ans.append(path)
            else:
                q.extend(path + [child] for child in graph[path[-1]])
        return ans
